File: main_project/ReplicaCluster/ClusterState/src/client/java/xce/clusterstate/ClusterStateSubscriber.java
package xce.clusterstate;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import Ice.Current;

import MyUtil.SbjState;
import MyUtil.SbjSubscriber;
import MyUtil.SbjTopic;
import MyUtil.SbjTopicI;

import xce.clusterstate.ClusterSeqState;

public class ClusterStateSubscriber extends SbjSubscriber{

	private String _endpoints;
	private long _interval;
	private Map<String,SbjTopic> _mapTopics=new ConcurrentHashMap<String,SbjTopic>();
	private String[] _clusterState;			//当前的ClusterState数组
	private String[] _oldState;				//更新前的ClusterState数组
	
	
	private static ConcurrentHashMap<String, ClusterStateSubscriber> _instances = new ConcurrentHashMap<String, ClusterStateSubscriber>();

	public static ClusterStateSubscriber getInstance(String instanceName) {
		if (!_instances.containsKey(instanceName)) {
			_instances.putIfAbsent(instanceName, new ClusterStateSubscriber());
		}
		return _instances.get(instanceName);
	}
	
	public ClusterStateSubscriber(){
		
	}
	
	
	void initialize(String endpoints,long interval){
		
		_endpoints=endpoints;
		_interval=interval;
		
		SbjTopicI topic=new SbjTopicI("ClusterState",this,_endpoints,_interval);
		
		Thread thr=new Thread(topic);
		
		thr.start();
		
		_mapTopics.put("ClusterState", topic);
		
	}
	
	public SbjState opera(SbjState state, Current __current) {
		
		if(state.getClass().getName().equals("xce.clusterstate.ClusterSeqState")){
			ClusterSeqState csState=new ClusterSeqState();
			csState=(ClusterSeqState)state;
			
			//备份现有ClusterState
			_oldState=_clusterState;
					
			// 对新赋值
			_clusterState=csState.states;
		}else{
			System.out.println("The type of state is not ClusterSeqState");
		}
		
		SbjState empty=new SbjState();
		return empty;
	}
	
	public String[] getClusterState(){
		return _clusterState;
	}
	
	public String[] getOldClusterState(){
		return _oldState;
	}
}

File: main_project/ReplicaCluster/ClusterState/src/client/cxx/ClusterStateSubscriber.cpp
#include "ClusterStateSubscriber.h"
#include "ClientInterface.h"
#include "ReplicaCluster/SubjectObserver/src/client/cxx/SbjTopicI.h"

using namespace std;
using namespace Ice;
using namespace MyUtil;
using namespace xce::clusterstate;
using namespace xce::serverstate;
using namespace xce::clustercontroller;

// ======= ClusterStateSubscriber =====================================
ClusterStateSubscriber::ClusterStateSubscriber(ClusterClient& client, std::string ctlEndpoints, int interval, const MyUtil::ChannelPtr& channel):
		_client(client),
		_ctlEndpoints(ctlEndpoints),
		_interval(interval),
    _channel(channel)	{
	init();
}

void ClusterStateSubscriber::init() {

	// 初始化 Channel和Communicator
	_ic= _channel->getCommunicator();

	// 注册SbjTopic
	SbjTopicIPtr topic = new SbjTopicI("ClusterState", this, _ctlEndpoints, _interval, _channel);
	_mapTopics.insert(pair<string, SbjTopicIPtr>("ClusterState", topic));

	// 启动时, 从controller获取服务信息
	getInitState();

	// 调用使用者的renew方法 
	// renew改为同步 -- 09.07.17 by zhanghan
	// TaskManager::instance().execute(new RenewTask(_client, _clusterState));
	_client.renew(_clusterState);
}

SbjStatePtr ClusterStateSubscriber::opera(const ::MyUtil::SbjStatePtr& sbjstate, const ::Ice::Current& current) {

	ClusterStateInfoPtr csStatePtr = ClusterStateInfoPtr::dynamicCast(sbjstate);
	
	ServerStateMap tmp(csStatePtr->states);
	IceUtil::Mutex::Lock lock(*this);	

	// 备份现有ClusterState
	_oldState.swap(_clusterState);	

	// 对新赋值
	_clusterState.swap(tmp);
	
	// 触发注册的ClusterClient的行为
	_client.renew(_clusterState);

	/* SbjStatePtr emptyptr = new SbjState;
	return emptyptr; */
	return sbjstate;
}

ServerStateMap& ClusterStateSubscriber::getClusterState() {

	return _clusterState;
}

ServerStateMap& ClusterStateSubscriber::getOldClusterState() {
	
	return _oldState;
}

void ClusterStateSubscriber::getInitState() {

	// 获取controller的prx
	Ice::ObjectPrx object = _channel->getCommunicator()->stringToProxy(string("M@")+_ctlEndpoints);
	ControllerManagerPrx prx = ControllerManagerPrx::uncheckedCast(object->ice_timeout(1000)->ice_twoway());

	try {
	       	_clusterState = prx->GetServerState()->states;
	} catch (Ice::Exception& e) {
		MCE_WARN("[getInitState] Ice::Exception : " << e.what());
	} catch (std::exception& e) {
		MCE_WARN("[getInitState] std::exception : " << e.what());
	} catch (...) {
		MCE_WARN("[getInitState] Exception");
	}
}

void RenewTask::handle() {
	_client.renew(_clusterState);
}

